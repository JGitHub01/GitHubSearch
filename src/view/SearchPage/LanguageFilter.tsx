const languages = [
  'JavaScript',
  'Java',
  'Python',
  'TypeScript',
  'C#',
  'C',
  'PhP',
  'Shell',
  'Ruby'].sort();
interface ILanguageFilterProps {
  language?: string
  onLanguageChange?: (language: string) => void
}
export default function LanguageFilter(props: ILanguageFilterProps) {
  const { language, onLanguageChange } = props;
  return (
    <div className='github-search__language' data-testid='search-language'>
      <label htmlFor='language'>Language:</label>
      <select name='language' data-testid='language-list'
        value={language}
        onChange={(e) => {
          onLanguageChange && onLanguageChange(e.target.value);
        }}
      >
        {languages.map((lan, idx) => {
          return (<option key={idx} value={lan.toLowerCase()}>{lan}</option>);
        })}
      </select>
    </div>
  );
}
