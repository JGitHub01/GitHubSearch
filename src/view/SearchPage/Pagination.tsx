import { useState, useEffect } from 'react';

interface IPaginationProps {
  page?: number
  totalPages?: number
  numPagingButtons?: number
  onPaging?: (page: number) => void
}
export default function Pagination({ page = 1, totalPages = 1, numPagingButtons = 5, onPaging }: IPaginationProps) {
  const [pageRange, handleLeftPaging, handleRightPaging] = usePaging(page, totalPages, numPagingButtons, onPaging);
  const pageButtons = [];
  for (let k = pageRange[0]; k <= pageRange[1]; k++) {
    const pageButtonClasses = `pagination-page${k === page ? ' pagination-page--active' : ''}`;
    pageButtons.push(
      <button className={pageButtonClasses}
        onClick={() => { onPaging && onPaging(k); }} key={k}
      >
        {k}
      </button>)
  }

  return (
    totalPages === 1 ? null :
      <div className='pagination' data-testid='pagination'>
        <button className='pagination-left'
          disabled={pageRange[0] <= 1}
          onClick={handleLeftPaging}
          data-testid='pagination-left'
        >{'<'}</button>
        {pageButtons}
        <button className='pagination-right'
          disabled={pageRange[1] >= totalPages}
          onClick={handleRightPaging}
          data-testid='pagination-right'
        >{'>'}</button>
      </div>
  );
}

function usePaging(page: number, totalPages: number, numPagingButtons: number, onPaging?: (page: number) => void):
  [Array<number>, () => void, () => void] {
  const [pageRange, setPageRange] = useState<number[]>(() => {
    return calPageRange(page, totalPages, numPagingButtons);
  });

  const rangeLeft = pageRange[0];
  const rangeRight = pageRange[1];
  useEffect(() => {
    const newPagesInRange = Math.min(numPagingButtons, totalPages);
    if ((newPagesInRange !== rangeRight - rangeLeft + 1)
      || page < rangeLeft || page > rangeRight) {
      setPageRange(calPageRange(page, totalPages, numPagingButtons));
    }
  }, [page, rangeLeft, rangeRight, totalPages, numPagingButtons]);

  const handleLeftPaging = () => {
    if (pageRange[0] <= 1) {
      return;
    }
    const shift = Math.max(1, Math.floor(numPagingButtons / 2));
    const newPage = pageRange[0] - 1;
    const left = Math.max(1, newPage - shift);
    const right = Math.min(left + numPagingButtons - 1, totalPages);
    setPageRange([left, right]);
    onPaging && onPaging(newPage);
  }
  const handleRightPaging = () => {
    if (pageRange[1] >= totalPages) {
      return;
    }
    const shift = Math.max(1, Math.floor(numPagingButtons / 2));
    const newPage = pageRange[1] + 1;
    const right = Math.min(totalPages, newPage + shift);
    const left = Math.max(right - numPagingButtons + 1, 1);
    setPageRange([left, right]);
    onPaging && onPaging(newPage);
  }

  return [pageRange, handleLeftPaging, handleRightPaging];
}

function calPageRange(page: number, totalPages: number, numPagingButtons: number) {
  const pagesInRange = Math.min(numPagingButtons, totalPages);
  if (page >= 1 && page <= pagesInRange) {
    return [1, pagesInRange];
  }

  const right = Math.min(totalPages, page + Math.floor((numPagingButtons - 1) / 2));
  return [right - pagesInRange + 1, right];
}
