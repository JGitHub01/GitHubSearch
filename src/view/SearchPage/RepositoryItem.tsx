interface IRepositoryItemProps {
  name?: string
  description?: string
}
export default function RepositoryItem(props: IRepositoryItemProps) {
  const { name, description } = props;
  return (
    <div className='repository' data-testid='repository-item'>
      <div className='repository-name'>
        <span className='repository-name__label'>Name:</span>
        <div className='repository-name__text'>{name}</div>
      </div>
      <div className='repository-description'>
        <span className='repository-description__label'>Description:</span>
        <div className='repository-description__text'>{description}</div>
      </div>
    </div>
  );
}
