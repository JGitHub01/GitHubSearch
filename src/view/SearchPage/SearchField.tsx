interface ISearchFieldProps {
  keyword?: string
  onKeywordChange?: (keyword: string) => void
}
export default function SearchField(props: ISearchFieldProps) {
  return (
    <input type='text' data-testid='search-field'
      className='github-search__keyword'
      value={props.keyword}
      onChange={(e) => {
        props.onKeywordChange && props.onKeywordChange(e.target.value);
      }}
    />
  );
}
