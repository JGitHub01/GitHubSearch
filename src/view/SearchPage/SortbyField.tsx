interface ISortbyFieldProps {
  sortby?: string
  onSortbyChange?: (sortby: string) => void
}
export default function SortbyField({ sortby, onSortbyChange }: ISortbyFieldProps) {
  return (
    <div className='github-search__sortby' data-testid='sortby-field'>
      <label htmlFor='sortby'>Sort by:</label>
      <select data-testid='sortby-list'
        name='sortby'
        value={sortby}
        onChange={e => {
          onSortbyChange && onSortbyChange(e.target.value);
        }}
      >
        <option value=''>Best match</option>
        <option value='stars'>Stars</option>
      </select>
    </div>
  );
}
