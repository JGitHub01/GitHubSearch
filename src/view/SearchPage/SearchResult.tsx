import { Link } from 'react-router-dom';
import { ISearchResult } from 'lib/models';
import RepositoryItem from './RepositoryItem';

interface ISearchResultProps extends ISearchResult {
  keyword?: string
  isSearching?: boolean
}
export default function SearchResult(props: ISearchResultProps) {
  const {
    keyword,
    isSearching = false,
    total_count,
    items,
    success = true,
    msg
  } = props;
  return (
    <div className='github-search-result' data-testid='search-result'>
      {(total_count > 0 && items != null) ?
        <div className='github-search-result__list' key='list'>
          {items.map((r, idx) =>
            <Link className='detail-link'
              to={{ pathname: '/detail', state: { ...r } }}
              key={idx}
            >
              <RepositoryItem name={r.name} description={r.description} />
            </Link>)
          }
        </div> :
        <div>
          {!success ? `Trouble search GitHub: ${msg}` :
            isSearching ? 'Searching GitHub repositoies ...' :
              keyword ? 'No repositories are found!' : 'Please enter a keyword!'}
        </div>
      }
    </div>
  );
}
