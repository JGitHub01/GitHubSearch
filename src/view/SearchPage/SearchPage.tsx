/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

import { useEffect, useState } from 'react';
import SortbyField from './SortbyField';
import SearchField from './SearchField';
import Pagination from './Pagination';
import LanguageFilter from './LanguageFilter';
import SearchResult from './SearchResult';
import { ISearchQuery, ISearchResult } from 'lib/models';
import { isScreenWidthAtLeast } from 'lib/responsive';

import './styles/search.scss';

interface ISearchPageProps {
  query: ISearchQuery
  result: ISearchResult
  onSearch: (query: ISearchQuery) => void
  isSearching?: boolean
}
const SearchPage = ({ query, result, onSearch, isSearching = false }: ISearchPageProps) => {
  const { language, perPage = 10, page, sort } = query;
  const { total_count } = result;
  const totalPages = Math.ceil(total_count / perPage);
  const numPagingButtons = usePagingButtons(totalPages);
  const [keyword, debouncedSearch] = useDebouncedSearch(query.keyword, onSearch);

  return (
    <div className='github-search' data-testid='search-page'>
      <h2>GitHub Repository Search</h2>
      <div className='github-search__toolbar'>
        <SortbyField
          sortby={sort}
          onSortbyChange={sort => onSearch({ sort, page: 1 })}
        />
        <LanguageFilter
          language={language}
          onLanguageChange={language => onSearch({ language, page: 1 })}
        />
      </div>
      <div className='github-search__sticky-header'>
        <SearchField
          keyword={keyword}
          onKeywordChange={keyword => debouncedSearch({ keyword, page: 1 })}
        />
        <div className='github-search__count'>
          <span>{`Total: ${total_count}`}</span>
          <span>{`${perPage}/page`}</span>
        </div>
      </div>
      <SearchResult
        keyword={keyword}
        isSearching={isSearching}
        {...result}
      />
      <div className='github-search__sticky-footer'>
        <Pagination page={page}
          totalPages={totalPages}
          numPagingButtons={numPagingButtons}
          onPaging={page => onSearch({ page })}
          key='pagination'
        />
      </div>
    </div>
  )
}

let searchDebouncer: NodeJS.Timeout;
function useDebouncedSearch(initKeyword: string | undefined, onSearch: (query: ISearchQuery) => void):
  [string | undefined, (query: ISearchQuery) => void] {
  const [keyword, setKeyword] = useState<string | undefined>(initKeyword);
  const debouncedSearch = (query: ISearchQuery) => {
    setKeyword(query.keyword);
    clearTimeout(searchDebouncer);
    searchDebouncer = setTimeout(() => {
      onSearch(query);
    }, 300);
  };
  return [keyword, debouncedSearch];
}
let resizeDebouncer: NodeJS.Timeout
function usePagingButtons(totalPages: number) {
  const [numPagingButtons, setNumPagingButtons] = useState<number>(() => {
    return isScreenWidthAtLeast('sm') ? 10 : 5;
  });

  useEffect(() => {
    const onResize = () => {
      clearTimeout(resizeDebouncer);
      resizeDebouncer = setTimeout(() => {
        const newNumPagingButtons = Math.min(totalPages, isScreenWidthAtLeast('sm') ? 10 : 5);
        if (newNumPagingButtons !== numPagingButtons) setNumPagingButtons(newNumPagingButtons);
      }, 300);
    }
    window.addEventListener('resize', onResize);

    return () => {
      window.removeEventListener('resize', onResize);
    }
  }, [totalPages, numPagingButtons]);
  return numPagingButtons;
}

export default SearchPage;
