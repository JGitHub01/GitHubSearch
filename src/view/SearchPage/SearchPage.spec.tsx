import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';
import SearchPage from './SearchPage';

jest.mock('lib/responsive', () => {
  return {
    isScreenWidthAtLeast: () => true
  };
});
describe('SearchPage', () => {
  const query = {
    keyword: "",
    language: 'javascript',
    perPage: 15
  }
  beforeAll(() => {
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });
  it('Should render the basic snapshot view', () => {
    const { getByTestId } = render(
      <Router>
        <SearchPage query={query}
          result={{ total_count: 0 }}
          onSearch={jest.fn()}
        />
      </Router>
    );
    expect(getByTestId('search-page')).toMatchSnapshot();
  });

  it('Should call onSearch when a keyword is entered', () => {
    jest.useFakeTimers();
    const callback = jest.fn();
    const { getByTestId } = render(
      <Router>
        <SearchPage query={query}
          result={{ total_count: 0 }}
          onSearch={callback}
        />
      </Router>
    );
    const searchField = getByTestId('search-field');
    userEvent.type(searchField, "tetris");
    expect(getByTestId('search-field')).toHaveValue('tetris');
    jest.runAllTimers();
    expect(callback).toBeCalledTimes(1);
    expect(callback).toBeCalledWith({keyword: "tetris", page: 1})
  });

  it('Should call onSearch when a new language is selected', () => {
    const callback = jest.fn();
    const { getByTestId, getByText } = render(
      <Router>
        <SearchPage query={query}
          result={{ total_count: 0 }}
          onSearch={callback}
        />
      </Router>
    );
    const lanList = getByTestId('language-list');
    const phpOption = getByText('PhP');
    userEvent.selectOptions(lanList, phpOption);
    expect(callback).toBeCalledWith({language: 'php', page: 1});
  });

    it('Should call onSearch when a new sort is selected', () => {
      const callback = jest.fn();
      const { getByTestId, getByText } = render(
        <Router>
          <SearchPage query={query}
            result={{ total_count: 0 }}
            onSearch={callback}
          />
        </Router>
      );
      const sortbyList = getByTestId('sortby-list');
      const starsOption = getByText('Stars');
      userEvent.selectOptions(sortbyList, starsOption);
      expect(callback).toBeCalledWith({sort: 'stars', page: 1});
    });
});
