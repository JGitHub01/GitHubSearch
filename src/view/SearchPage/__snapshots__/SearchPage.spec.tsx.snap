// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage Should render the basic snapshot view 1`] = `
<div
  class="github-search"
  data-testid="search-page"
>
  <h2>
    GitHub Repository Search
  </h2>
  <div
    class="github-search__toolbar"
  >
    <div
      class="github-search__sortby"
      data-testid="sortby-field"
    >
      <label
        for="sortby"
      >
        Sort by:
      </label>
      <select
        data-testid="sortby-list"
        name="sortby"
      >
        <option
          value=""
        >
          Best match
        </option>
        <option
          value="stars"
        >
          Stars
        </option>
      </select>
    </div>
    <div
      class="github-search__language"
      data-testid="search-language"
    >
      <label
        for="language"
      >
        Language:
      </label>
      <select
        data-testid="language-list"
        name="language"
      >
        <option
          value="c"
        >
          C
        </option>
        <option
          value="c#"
        >
          C#
        </option>
        <option
          value="java"
        >
          Java
        </option>
        <option
          value="javascript"
        >
          JavaScript
        </option>
        <option
          value="php"
        >
          PhP
        </option>
        <option
          value="python"
        >
          Python
        </option>
        <option
          value="ruby"
        >
          Ruby
        </option>
        <option
          value="shell"
        >
          Shell
        </option>
        <option
          value="typescript"
        >
          TypeScript
        </option>
      </select>
    </div>
  </div>
  <div
    class="github-search__sticky-header"
  >
    <input
      class="github-search__keyword"
      data-testid="search-field"
      type="text"
      value=""
    />
    <div
      class="github-search__count"
    >
      <span>
        Total: 0
      </span>
      <span>
        15/page
      </span>
    </div>
  </div>
  <div
    class="github-search-result"
    data-testid="search-result"
  >
    <div>
      Please enter a keyword!
    </div>
  </div>
  <div
    class="github-search__sticky-footer"
  >
    <div
      class="pagination"
      data-testid="pagination"
    >
      <button
        class="pagination-left"
        data-testid="pagination-left"
        disabled=""
      >
        &lt;
      </button>
      <button
        class="pagination-right"
        data-testid="pagination-right"
        disabled=""
      >
        &gt;
      </button>
    </div>
  </div>
</div>
`;
